src/Service/FormDataModelFactory.php

```diff
@@ -0,0 +1,181 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Atoolo\Form\Service;
+
+use Atoolo\Form\Dto\FormDefinition;
+use Atoolo\Form\Dto\UISchema\Control;
+use Atoolo\Form\Dto\UISchema\Layout;
+
+class FormDataModelFactory implements FromReaderHandler
+{
+    /**
+     * @var array<EmailMessageModelItem>
+     */
+    private array $items = [];
+
+    /**
+     * @var array<EmailMessageModelItem|array<EmailMessageModelItem>>
+     */
+    private array $stack = [];
+
+    private bool $includeEmptyFields = false;
+
+    public function __construct(
+        private readonly DataUrlParser $dataUrlParser,
+    ) {}
+
+    /**
+     * @param array<string,mixed> $data
+     * @return array<EmailMessageModelItem>
+     */
+    public function create(FormDefinition $definition, array $data, bool $includeEmptyFields): array
+    {
+        $this->stack = [];
+        $this->items = [];
+        $this->includeEmptyFields = $includeEmptyFields;
+        $reader = new FormReader($definition, $data, $this);
+        $reader->read();
+
+        return $this->items;
+    }
```

Comment:
This is about displaying the unfilled fields in the HTML email. This is a setting that an editor can set for each form. Since `FormDataModelFactory` is a service that is created via DI, but is used for all forms, this cannot be defined in the `FormDataModelFactory`.

A user would only ever use the service and not create an instance of the `FormDataModelFactory` themselves.

```php
    public function __construct(
        private readonly FormDataModelFactory $formDataModelFactory,
    ) {}

    public function doSomething()
    {
        $this->formDataModelFactory->create($definition, $data, includeEmptyFields: true);
    }
```